docs/enhancements/cancel-day-2-host-installation.md

```diff
@@ -0,0 +1,137 @@
+
+- Hosts that are installed in a day-1 cluster is out of scope of this enhancement
+- Cluster installation cancellation is out of scope
+- The way reclaim/unbind works should not be affected by these changes
+- This enhancement is focused solely on Assisted-Service offerings using Kubernetes API (SaaS should not be affected by this)
+
+## Proposal
+
+Allow cancelling an installing day-2 host through Assisted-Service's Kubernetes API.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, when I create a hosted cluster using Hypershift, I'd like to bind and install worker hosts into my cluster. 
+However, if I want to remove the host from the cluster before they finish installing, I should be able to do so and reuse the host for another cluster.
+
+#### Story 2
+
+As a cluster admin using Hypershift, I want to use the autoscaler.
+When the scaling happens and hosts are bound/unbound from the cluster, the hosts should bind and unbind 
+easily from the cluster.
+
+#### Story 3
+
+As a cluster admin, if I install a day-2 host to a cluster, and I see the host installation fails from
+inside the host. I should be able to cancel the installation once it starts and quickly reset the host.
+
+### Implementation Details/Notes/Constraints [optional]
+
+There is an already existing cancel host installation API which maps to the unbind and reclaim stages.
+This is currently only being used when a day-1 cluster installation is cancelled. 
+
+#### Late-binding case
+
+For late-binding, an Agent can be reused by another cluster so in order to detect if the Agent
+becomes unbound from a cluster while installing, we'll check if the cluster reference on the
+Agent has been removed or changed. Currently, the existing Agent webhook blocks this change
+if the Agent is in an installing stage. For this proposal, we can remove that webhook.
+
+Once the Agent is detected as unbound during installation, then we can use the existing cancellation
+API to cancel the host's installation.
+
+#### Early-binding case
+
+For early-binding, an Agent is directly attached to a cluster when it's discovered. In order to
+detect cancellation, the Agent would need to be deleted while it's installing. 
+
+Currently, the webhook allows an Agent to be deleted while it's installing. 
+Once an Agent is deleted, it is simply removed from the database.
+
+This flow would need to be modified to detect if the Agent is in the middle of installation,
+and if it is, then send the cancel install host command to the Agent before completely allowing
+it to be removed. The Agent should move to the cancel installation stage until the host command
+returns successfully. If the command is not successful, then the Agent will move to an Error state.
+
+There may be some cases where a user might not care if the host is cancelled if they're deleting
+the Agent, so there should also be a hook that allows the host to be force deleted without waiting
+for the cancel command to complete successfully.
```

Comment:
What is the value of canceling the installation for a host if the host is going to be deleted in any case.
Do we do something on the host itself?

What happens today when you're installing a day2 host with early binding and delete the agent?
Does the host continue to install? Do we just not approve the CSR?

Reply:
That's a good point, if the user wanted to delete the host then it shouldn't be gated.

I tested out deleting an Agent when the host went into rebooting and saw that the host continued to try and install, but it failed to start kubelet. Didn't get to the point of approving CSRs, though. 

At any rate, I decided to limit the scope of this enhancement to just late-binding hosts to keep it simple